### PR TITLE
Integrate RoTTA Method into TTAB Framework

### LINKED ISSUE #11: Consider integrating RoTTA
Hello, I noticed a new test-time adaptation method called Robust Test-Time Adaptation (RoTTA) proposed in this paper
https://arxiv.org/abs/2303.13899.
According to the paper, RoTTA achieves state-of-the-art performance and outperforms some existing methods like TENT, CoTTA, NOTE, etc.

I tried integrating RoTTA into the codebase myself. However, when I tested it, the accuracy rate I got was much lower than reported in the original RoTTA paper . I tried to debug it but couldn't get the accuracy up. I wonder if you have considered integrating RoTTA? As the original authors of the codebase, you may have better insights on how to properly integrate RoTTA.

Please let me know if you would be interested in integrating RoTTA. I'm happy to provide more details on the issues I encountered. I think RoTTA could further improve the codebase's adaptation capabilities.

Thank you for your attention! Please let me know if you need any other information from me.

### PR DESCRIPTION
**Pull request by helium777**
## New Features
- Implemented the Practical Test-Time Adaptation (PTTA) setting as described in the paper, within `scenarios/default_scenarios.py`.
- Integrated the RoTTA (Robust Test-Time Adaptation) method in `model_adaptation/rotta.py`.
- Added functionality for loading the pretrained WideResNet model used in RoTTA, located in `loads/`. 
   - Note: The checkpoint must be manually downloaded from the Google Drive link provided by the RobustBench repository. Additionally, in the pickle file from RobustBench, the key for model weights is labeled 'state_dict' instead of 'model', necessitating a manual conversion for compatibility.
- Provided a minimal working example in `notebooks/rotta_cifar10.ipynb`.

## Considerations
- It's important to note that the implementation of WideResNet in TTAB differs from that in RobustBench, which RoTTA uses. 
   - For instance, in RobustBench, all convolutional layers in WideResNet have a bias, unlike in TTAB.
   - Consequently, the `loads/models` folder now contains two versions of WideresNet: `WideResNet` class in `wideresnet.py`, and `WideResNetRB` class in `wideresnet_robustbench.py`.

## Reproduction Results
I've replicated RoTTA's results on CIFAR-10 to CIFAR-10-C to evaluate its performance under the TTAB framework. It is important to note that I need guidance on calculating the error rate for each domain within the TTAB framework. Therefore, the table includes only partial results:

|              | motion | snow  | fog   | shot  | defocus | contrast | zoom  | brightness | frost | elastic | glass | gaussian | pixelate | jpeg  | impulse | avg.  |
| ------------ | ------ | ----- | ----- | ----- | ------- | -------- | ----- | ---------- | ----- | ------- | ----- | -------- | -------- | ----- | ------- | ----- |
| Reported     | 18.1   | 21.3  | 18.8  | 33.6  | 23.6    | 16.5     | 15.1  | 11.2       | 21.9  | 30.7    | 39.6  | 26.8     | 33.7     | 27.8  | 39.5    | 25.2  |
| Official Repo | 19.72  | 21.59 | 17.66 | 31.58 | 19.37   | 18.50    | 16.07 | 9.38       | 19.97 | 27.52   | 41.36 | 30.08    | 32.06    | 26.18 | 36.47   | 24.50 |
| TTAB         |        |       |       |       |         |          |       |            |       |         |       |          |          |       |         | 31.82 |

## Request for Assistance

I would appreciate guidance on how to properly compute the error rates for each domain within the TTAB framework to complete the above table and verify the implementation's effectiveness.

### CODE REVIEW

File: ttab/loads/define_model.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 import timm
 import torch
+from ttab.loads.models.wideresnet_robust_bench import WideResNetRB
```
Comment (MarcellusZhao) [inline ttab/loads/define_model.py]: It's good to test the correctness of implementation using the same model as the paper does, but I think there is no need to integrate this particular model into our benchmark, especially when it has to build another independent model class.

File: ttab/loads/define_model.py [hunk 2]
```diff
@@ -115,7 +116,17 @@ def define_model(config):
         return timm.create_model(config.model_name, pretrained=True)
 
     # use built-in models and local checkpoints.
-    if "wideresnet" in config.model_name:
+    if "wideresnetrb" in config.model_name:
```
Comment (MarcellusZhao) [inline ttab/loads/define_model.py]: I think we'd better remove this too.

File: ttab/loads/models/wideresnet_robust_bench.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        layers = []
+        for i in range(int(nb_layers)):
+            layers.append(block(i == 0 and in_planes or out_planes, out_planes, i == 0 and stride or 1, dropRate))
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+        return self.layer(x)
+
+
+class WideResNetRB(nn.Module):
```
Comment (MarcellusZhao) [inline ttab/loads/models/wideresnet_robust_bench.py]: Remove this class as well.

File: ttab/model_adaptation/rotta.py [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+                tmp_data.append(item.data)
+                tmp_age.append(item.age)
+
+        tmp_age = [x / self.capacity for x in tmp_age]
+
+        return tmp_data, tmp_age
+
+
+# ========== bn_layers ==========
+class MomentumBN(nn.Module):
```
Comment (MarcellusZhao) [inline ttab/model_adaptation/rotta.py]: i recommend moving these normalization layers to `model_adaptation/utils.py`.

File: ttab/model_adaptation/rotta.py [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+
+        x = (x - mean) / torch.sqrt(var + self.eps)
+        weight = self.weight.view(1, -1, 1, 1)
+        bias = self.bias.view(1, -1, 1, 1)
+
+        return x * weight + bias
+
+
+# ========== custom_transforms ==========
+def get_tta_transforms(cfg, gaussian_std: float = 0.005, soft=False):
```
Comment (MarcellusZhao) [inline ttab/model_adaptation/rotta.py]: Why do we need this? it is created for test or something else?

File: ttab/model_adaptation/rotta.py [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+            transforms.CenterCrop(size=n_pixels),
+            transforms.RandomHorizontalFlip(p=p_hflip),
+            GaussianNoise(0, gaussian_std),
+            Clip(clip_min, clip_max),
+        ]
+    )
+    return tta_transforms
+
+
+class GaussianNoise(torch.nn.Module):
```
Comment (MarcellusZhao) [inline ttab/model_adaptation/rotta.py:451]: This is irrelevant to this TTA method, right? So please move this to somewhere else or remove it.

File: ttab/model_adaptation/rotta.py [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+        noise = noise.to(img.device)
+        return img + noise
+
+    def __repr__(self):
+        return self.__class__.__name__ + "(mean={0}, std={1})".format(
+            self.mean, self.std
+        )
+
+
+class Clip(torch.nn.Module):
```
Comment (MarcellusZhao) [inline ttab/model_adaptation/rotta.py:468]: This is irrelevant as well.

File: ttab/model_adaptation/rotta.py [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+    def forward(self, img):
+        return torch.clip(img, self.min_val, self.max_val)
+
+    def __repr__(self):
+        return self.__class__.__name__ + "(min_val={0}, max_val={1})".format(
+            self.min_val, self.max_val
+        )
+
+
+class ColorJitterPro(ColorJitter):
```
Comment (MarcellusZhao) [inline ttab/model_adaptation/rotta.py:483]: This is irrelevant as well.

File: ttab/model_adaptation/rotta.py [hunk 9]
```diff
@@ -0,0 +1,664 @@
...
+        format_string = self.__class__.__name__ + "("
+        format_string += "brightness={0}".format(self.brightness)
+        format_string += ", contrast={0}".format(self.contrast)
+        format_string += ", saturation={0}".format(self.saturation)
+        format_string += ", hue={0})".format(self.hue)
+        format_string += ", gamma={0})".format(self.gamma)
+        return format_string
+
+
+# ========= utils ==========
```
Comment (MarcellusZhao) [inline ttab/model_adaptation/rotta.py]: Move irrelevant functions to `utils.py` or make it internal for TTA method class if you have to.

File: ttab/scenarios/default_scenarios.py [hunk 10]
```diff
@@ -36,4 +36,232 @@
             intra_domain_shuffle=True,
         ),
     ),
+    "PTTA_cifar10": Scenario(
```
Comment (MarcellusZhao) [inline ttab/scenarios/default_scenarios.py:39]: Too much unnecessary repetition. We'd better develop some new code to support this particular setting.

**Review by MarcellusZhao — COMMENTED** [review]
Excellent work! Let's fix comments one by one.

File: ttab/model_adaptation/rotta.py (hunk 5 above)
Comment (helium777) [reply ttab/model_adaptation/rotta.py]: It is used to generate the 'augmented view' for the EMA model (the teacher model).

File: ttab/model_adaptation/rotta.py (hunk 6 above)
Comment (helium777) [reply ttab/model_adaptation/rotta.py:451]: It is used in the `get_tta_transforms` function, as well as the functions below.

File: ttab/model_adaptation/rotta.py (hunk 5 above)
Comment (MarcellusZhao) [reply ttab/model_adaptation/rotta.py]: It sounds like a method-specific operation. Probably we can move this into the RoTTA class. What do you think?

File: ttab/model_adaptation/rotta.py (hunk 6 above)
Comment (MarcellusZhao) [reply ttab/model_adaptation/rotta.py:451]: I see. It seems these modules are transferable to other augmentation-based TTA methods. Then it makes sense to move these things to other places.

**Comment by helium777** [comment]
## Changes

- The implementation of the WideResNet from RobustBench has been removed.
- Some functions in `rotta.py` have been moved to either `utils.py` or the `RoTTA` class.

## Experiment

I have conducted a quick test for RoTTA using ResNet26 and obtained an error rate of 43.16%. However, I'm unsure if this is a reasonable result.
